File: Max Dot Product of Two Subsequences.py
class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        n1, n2 = len(nums1), len(nums2)
        dp = [[0] * (n2 + 1) for _ in range(n1 + 1)]
        ans = nums1[0] * nums2[0]

        for i in range(1, n1 + 1):
            for j in range(1, n2 + 1):
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1] + nums1[i - 1] * nums2[j - 1])
                ans = max(ans, dp[i - 1][j - 1] + nums1[i - 1] * nums2[j - 1])

        return ans

File: Max Dot Product of Two Subsequences.cpp
class Solution {
public:
    int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        int n1 = nums1.size();
        int n2 = nums2.size();
        vector<vector<int>> dp(n1+1, vector<int>(n2+1, 0));
        int maxx{INT_MIN};
        for (int i{0}; i < n1; ++i) {
            for (int j{0}; j < n2; ++j) {
                maxx = max(maxx, nums1[i]*nums2[j]);
                dp[i+1][j+1] = max(dp[i+1][j+1], dp[i][j] + nums1[i]*nums2[j]);
                dp[i][j+1] = max(dp[i][j+1], dp[i][j]);
                dp[i+1][j] = max(dp[i+1][j], dp[i][j]);
            }
        }

        int maxx_minn{maxx};
        for (int i{0}; i < n1+1; ++i) {
            maxx = max(maxx, dp[i][n2]);
        }
        for (int j{0}; j < n2+1; ++j) {
            maxx = max(maxx, dp[n1][j]);
        }
        return (maxx == 0) ? maxx_minn : maxx;
    }
};
